Adds delegator for AdapterAwareInterface

froschdesign opened the pull request:
|    Q          |   A
|-------------- | ------
| New Feature   | yes

## Description

The delegator allows the setting of an adapter for a class which implements the `Laminas\Db\Adapter\AdapterAwareInterface`.

## Example

A concrete example are the database validators [`RecordExists` and `NoRecordExists`](https://docs.laminas.dev/laminas-validator/validators/db/). Is a database adapter registered and the delegator is present in the validator plugin manager then following configuration can be used in a form:

```php
public function getInputFilterSpecification() : array
{
    return [
        [
            'name'       => 'form-element',
            'validators' => [
                [
                    'name'    => RecordExists::class,
                    'options' => [
                        'table' => 'foo',
                        'field' => 'bar',
                    ],
                ],
            ],
        ],
    ];
}
```
It is not necessary to create a factory or set the database adapter to form or validator manually.

(The next step is to extend the validator plugin manager.)

---

_The increase of the PHP version is needed._

froschdesign committed:
Adds delegator for AdapterAwareInterface

Signed-off-by: Frank Brückner <dev@froschdesignstudio.de>

weierophinney reviewed (changes requested):
I can't believe I'm asking this of you, @froschdesign ... but can you add documentation around the new delegator with this patch, please? :smile:

froschdesign committed:
Adds documentation for AdapterAwareTrait

Signed-off-by: Frank Brückner <dev@froschdesignstudio.de>

froschdesign commented:
@weierophinney 
Documentation is added! 😃

froschdesign committed:
Extends description for delegator and database adapter

Signed-off-by: Frank Brückner <dev@froschdesignstudio.de>

froschdesign committed:
Updates code example

Signed-off-by: Frank Brückner <dev@froschdesignstudio.de>

weierophinney committed:
docs: remove trailing space on line

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix CS issue in AdapterServiceDelegatorTest

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: add ProphecyTrait to AdapterServiceDelegatorTest

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: bump laminas-servicemanager minimum version to 3.3

First version that required container-interop 1.2, which ensures we can use PSR-11 interfaces as well.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
